### DISCUSSION TITLE
Why Initialize method on ViewModels is in the Main Thread?

### DISCUSSION THREAD

**Discussion by DavidMarquezF** [MvvmCross/MvvmCross#4859] [Q&A] [answered]
This might be a stupid question, but the other day I was tracing my app for performance and realised that the Initialize method was running on the main thread (not of the first ViewModel loaded).

Just as a quick test I checked what thread it was running on:
![image](https://github.com/user-attachments/assets/1026f7ec-5fae-4638-9224-74564b5d8b89)

I thought, according to docs that I can do heavy loading here and that I should not worry about blocking the main thread. Does this mean that the Initialize function should run on a different thread? Or does it mean that I can make code that safely runs on another thread inside of the Initialize function?

Thank you very much

**Comment by entdark** [comment, marked as answer]
Good thing you noticed. I noticed it too in some cases.

The `Initialize` task gets created here: https://github.com/MvvmCross/MvvmCross/blob/release/9.2.0/MvvmCross/ViewModels/MvxDefaultViewModelLocator.cs#L170
And then it gets awaited here: https://github.com/MvvmCross/MvvmCross/blob/release/9.2.0/MvvmCross/Navigation/MvxNavigationService.cs#L301

As you can see there is `ConfigureAwait(false)` that supposes to switch the context. But it only gets switched on inner `await`. And you have `await CreateFilters()` that gets awaited. But if there is not any inner `await` inside your `CreateFilters` or there are some long running operations before any `await` then you are still on the main thread. The same goes for `await LoadList()`.

In my case it happened when internet connection was bad on Android and the UI got frozen because in the implementation of `HttpClient.GetAsync()` there was some long running operation before inner `await`.

In general, I assume that `ConfigureAwait(false)` is a bad practice that spreaded much in the C# world, and in MVVM case it is bad to base initialization implementation on it.

My solution was simple. And it is the only right solution if you want to initialize on background:
```csharp
public override Task Initialize() {
	Task.Run(BackgroundInitialize);
	return Task.CompletedTask;
}

protected virtual Task BackgroundInitialize() {
	return Task.CompletedTask;
}
```

Here is some good reading about `ConfigureAwait(false)` problem with samples on GitHub: https://learn.microsoft.com/en-us/archive/blogs/benwilli/an-alternative-to-configureawaitfalse-everywhere

It is still better but it won't save from such a case:
![Screenshot 2024-02-29 at 18 08 59-sryor9gtqffkxgda3ydrq6oa4y](https://github.com/user-attachments/assets/f6114c6d-bf45-4028-ab3f-3d377367b5dd)
If there is not any `await` or `await` after a long operation (the `for` loop) then it will fail and lag on the main thread until the end of the function.

So I will repeat that the only right way to initialize is to use `Task.Run()`. You can make your own tests.

Good luck. :)

P.S.: that hack will work but it's bullshit IMO:
![Screenshot 2024-02-29 at 18 11 11-jreiso4s3t8b8xrompnquexedy](https://github.com/user-attachments/assets/ff153c09-0fc3-4273-8404-f6abad324aef)

**Reply by DavidMarquezF** [reply]
Thank you so much for the detailed response! It's indeed a bit confusing. I was wondering why not the following?
```
public override Task Initialize() {
	return Task.Run(BackgroundInitialize);
}

protected virtual Task BackgroundInitialize() {
	return Task.CompletedTask;
}
```
With your example, initialize would finish instantly, therefore taking away the possibility of using the Notify Task `InitializeTask`. Would that still work?

**Reply by entdark** [reply]
Because the returned `Task` will have to wait the job that is passed to `Task.Run()`, not matter it is `Action` or `Func<Task>`, the funcion `Task` will be awaited anyways. And if there is any long-running inner jobs inside the function before any `await` then the issue remains. We have to run the job and move on.

**Reply by entdark** [reply]
And yes, the VM default initialization is done instantly in that case. I see no issues in running another background `Task` and initialize data there. If you need to initialize some variables then it's better be done in the constructor or in `Prepare` method. But if those variables require some computation than it's better leave in background. I usually do it like
```csharp
protected override async Task BackgroundInitialize()
{
	IsLoading = true;
	await LoadData();
	IsLoading = false;
}
```

**Reply by DavidMarquezF** [reply]
Okay thank you very much!
